Cython/Compiler/Main.py

```diff
@@ -472,32 +472,66 @@ def create_default_resultobj(compilation_source, options):
     result.embedded_metadata = options.embedded_metadata
     return result
 
-
-def run_pipeline(source, options, full_module_name=None, context=None):
-    from . import Pipeline
-
-    source_ext = os.path.splitext(source)[1]
-    options.configure_language_defaults(source_ext[1:])  # py/pyx
-    if context is None:
-        context = Context.from_options(options)
-
-    # Set up source object
+def setup_source_object(source, source_ext, full_module_name, options, context):
     cwd = os.getcwd()
     abs_path = os.path.abspath(source)
+
     full_module_name = full_module_name or context.extract_module_name(source, options)
     full_module_name = EncodedString(full_module_name)
 
     Utils.raise_error_if_module_name_forbidden(full_module_name)
-
     if options.relative_path_in_code_position_comments:
         rel_path = full_module_name.replace('.', os.sep) + source_ext
         if not abs_path.endswith(rel_path):
             rel_path = source  # safety measure to prevent printing incorrect paths
     else:
         rel_path = abs_path
     source_desc = FileSourceDescriptor(abs_path, rel_path)
-    source = CompilationSource(source_desc, full_module_name, cwd)
+    return CompilationSource(source_desc, full_module_name, cwd)
+
+def run_cached_pipeline(source, options, full_module_name=None, context=None, cache=None, fingerprint=None):
+
+    if context is None:
+        context = Context.from_options(options)
+
+    if cache:
+        if not fingerprint:
+            fingerprint = get_fingerprint(cache, source, options)
+        cwd = os.getcwd()
+        output_filename = get_output_filename(source, cwd, options)
+        cached = cache.lookup_cache(output_filename, fingerprint)
+    else:
+        cached = False
+    if cache and cached:
+        if options.verbose:
+            sys.stderr.write(f'Found compiled {os.path.basename(output_filename)} in cache.\n')
+        cache.load_from_cache(output_filename, cached)
+
+        source_ext = os.path.splitext(source)[1]
+        options.configure_language_defaults(source_ext[1:])  # py/pyx
+
+        source = setup_source_object(source, source_ext, full_module_name, options, context)
+        # Set up result object
+        return create_default_resultobj(source, options)
+    else:
+        result = run_pipeline(source, options, full_module_name, context)
+        if cache and fingerprint:
+            cache.store_to_cache(output_filename, fingerprint, result)
+    return result
```

Comment:
Let's make it clear that we consider the "build the module" case the normal one:
```suggestion
        return create_default_resultobj(source, options)

    result = run_pipeline(source, options, full_module_name, context)
    if cache and fingerprint:
        cache.store_to_cache(output_filename, fingerprint, result)
    return result
```